### PR TITLE
Re-enable cuda texture object based access in regions kernel

### PR DESCRIPTION
**Pull request by 9prady9**
There is no improvement except for some specific sizes. Note that I have run the benchmark on only 2^n sizes.

|Size | TextureObject | Plain Old Memory |
|---|----|---|
128 x  128 | 0.00247927 | 0.00299974
256 x  256 | 0.00671363 | 0.006602
384 x  384 | 0.00620648 | 0.00795144
512 x  512 | 0.00989444 | 0.0123784
640 x  640 | 0.0136365 | 0.0155043
768 x  768 | 0.0186786 | 0.0196094
896 x  896 | 0.0253983 | 0.0266321
1024 x 1024 | 0.0322667 | 0.0337688
1152 x 1152 | 0.0411638 | 0.0431613
1280 x 1280 | 0.0529367 | 0.0560853
1408 x 1408 | 0.0612112 | 0.0643147
1536 x 1536 | 0.0872647 | 0.0769113
1664 x 1664 | 0.105343 | 0.0999272
1792 x 1792 | 0.12006 | 0.112325
1920 x 1920 | 0.112999 | 0.118628
2048 x 2048 | 0.129985 | 0.136978

chart

[skip arrayfire ci]

### CODE REVIEW

File: src/backend/cuda/kernel/regions.hpp [hunk 1]
```diff
@@ -35,21 +36,24 @@ __device__ static int continue_flag = 1;
...
-               cuda::Param<T> equiv_map,
-               cudaTextureObject_t tex)
+static inline __device__
+T fetch(const int n, cuda::Param<T> equiv_map, cudaTextureObject_t tex)
 {
-// FIXME: Enable capability >= 3.0
-//#if (__CUDA_ARCH__ >= 300)
-#if 0
-    // Kepler bindless texture objects
+#if __CUDA_ARCH__ >= 300
```
Comment (pavanky) [inline src/backend/cuda/kernel/regions.hpp]: Can you actually drop this ? We don't support pre 3.0 compute anymore.

File: src/backend/cuda/regions.cu [hunk 2]
```diff
@@ -29,20 +29,30 @@ Array<T>  regions(const Array<char> &in, af_connectivity connectivity)
...
-    //    cudaTextureDesc texDesc;
-    //    memset(&texDesc, 0, sizeof(texDesc));
-    //    texDesc.readMode = cudaReadModeElementType;
-    //    CUDA_CHECK(cudaCreateTextureObject(&tex, &resDesc, &texDesc, NULL));
-    //}
+
+    cudaDeviceProp prop = getDeviceProp(getActiveDeviceId());
+
+    //Use texture objects with compute 3.0 or higher
+    if (prop.major>=3 && prop.minor>=0 && !std::is_same<T,double>::value) {
```
Comment (pavanky) [inline src/backend/cuda/regions.cu]: Same as above.

**Comment by 9prady9** [comment]
@arrayfire/core-devel This is ready for merge.

File: examples/benchmarks/regions.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            printf("Avg. Time Taken:  %g \n", timeit(benchmarkRegions));
+            fflush(stdout);
+        }
+    } catch (af::exception& e) {
+        fprintf(stderr, "%s\n", e.what());
+        throw;
+    }
+
+    return 0;
+}
```
Comment (pavanky) [inline examples/benchmarks/regions.cpp]: Do we need to check this in?

File: src/backend/cuda/kernel/regions.hpp [hunk 4]
```diff
@@ -23,8 +23,9 @@
 #include <thrust/sort.h>
 #include <thrust/transform_scan.h>
 
-#if __CUDACC__
+#include <type_traits>
```
Comment (pavanky) [inline src/backend/cuda/kernel/regions.hpp]: Where is this used ?

File: src/backend/cuda/kernel/regions.hpp [hunk 5]
```diff
@@ -23,8 +23,9 @@
...
 
+#if __CUDACC__
```
Comment (pavanky) [inline src/backend/cuda/kernel/regions.hpp]: Is this necessary?

File: src/backend/cuda/regions.cu [hunk 6]
```diff
@@ -29,20 +29,28 @@ Array<T>  regions(const Array<char> &in, af_connectivity connectivity)
...
+            resDesc.res.linear.desc.f = cudaChannelFormatKindUnsigned;
+        else
+            resDesc.res.linear.desc.f = cudaChannelFormatKindFloat;
+
+        resDesc.res.linear.desc.x = sizeof(T)*8; // bits per channel
+        resDesc.res.linear.sizeInBytes = dims[0] * dims[1] * sizeof(T);
+        cudaTextureDesc texDesc;
+        memset(&texDesc, 0, sizeof(texDesc));
+        texDesc.readMode = cudaReadModeElementType;
+        CUDA_CHECK(cudaCreateTextureObject(&tex, &resDesc, &texDesc, NULL));
```
Comment (pavanky) [inline src/backend/cuda/regions.cu]: You are not releasing the texture.

File: src/backend/cuda/kernel/regions.hpp (hunk 4 above)
Comment (9prady9) [inline src/backend/cuda/kernel/regions.hpp]: Not anymore, it was used in first commit but I removed it now. Will remove it.

File: src/backend/cuda/kernel/regions.hpp (hunk 5 above)
Comment (9prady9) [inline src/backend/cuda/kernel/regions.hpp]: Not really. It wasn't added in this change, so I didn't notice this.

**Comment by 9prady9** [comment]
@pavanky The the performance looks good. 
chart 1

**Comment by 9prady9** [comment]
@pavanky If there is no additional feedback, I think this is ready for merge.

**Comment by pavanky** [comment]
@9prady9 can you close this PR and create a new one targeting v3.5 by cherry picking the appropriate commits?

**Comment by 9prady9** [comment]
@umar456 / @mlloreda  please kill the ci jobs for this task.

**Comment by pavanky** [comment]
@9prady9 Nevermind, the problem is with the base branch. I am fixing v3.5

**Comment by pavanky** [comment]
@9prady9 fixed now.

**Comment by umar456** [comment]
The performance of this function depends on the input image. What input image did you use for this function? Can you test the performance where more iterations are required? How does it compare to the performance where there are only few iterations?

Also, I am having trouble seeing how this access pattern takes advantage of the texture objects. Looks like you are loading into the shared memory so you are not taking advantage of the additional texture cache. You are using linear textures so you aren't taking advantage of the spatial locality accesses of Texture caches. I don't see a huge advantage to using this approach. Maybe it will be more pronounced in inputs with more iterations.

**Comment by pavanky** [comment]
Also we should probably be using cudaArray instead of cudaTextureObject.

**Comment by 9prady9** [comment]
@umar456 There is an advantage to using texture objects for [this section of code](https://github.com/arrayfire/arrayfire/blob/0db608154d5e5a5ed30ab74d4e6026bf8d12237f/src/backend/cuda/kernel/regions.hpp#L163-L244) because it is doing 2D spatial access(reading more than one element per thread - I am not talking about per thread load) and textures surely have an advantage compared to global memory in that aspect.

Does having the additional layer of shared memory reduce the performance ? It depends on whether shared memory has less latency compared to texture access. I think shared memory latency is better compared to texture access, but I am not certain of it.

I had some benchmark sample using randomized input commited earlier, but removed it later and did a force push. Do you have any suggestion for the input you were talking about - the one that will cause regions to run more iterations ?

@pavanky Is cudaArray similar to texture in performance ? I have seen couple of [example](https://github.com/odlgroup/odl-cpp-utils/blob/master/cuda/texture.h)s online where cudaArray_t is again wrapped around with cudaTextureObject_t. Not sure if it is equivalent in performance with textures.

**Comment by pavanky** [comment]
The benefits coming from texture memory seem to be fairly minimal at the moment.

**Comment by 9prady9** [comment]
@pavanky @umar456 I have cleaned up more of it. I have tested the changes with two actual images(using timeit) and the improvement is around 20+%.

One of them has lot of smaller regions(image with lot of text) and the other has huge but fewer regions.
- 21% improvement for image with lot of individual components but smaller regions - took **8 iterations**
- 26% improvement for image with less individual components but big regions - took **22 iterations**.

wtext
sciie

**Pull request merged**